test/e2e/operator_test.go

```diff
@@ -670,3 +672,112 @@ func TestHostNetworkEndpointPublishingStrategy(t *testing.T) {
 		t.Fatalf("WARNING: cloud resources may have been leaked! failed to delete ingresscontroller %s: %v", name, err)
 	}
 }
+
+// TestInternalLoadBalancer creates an ingresscontroller with the
+// "LoadBalancerService" endpoint publishing strategy type with scope set to
+// "Internal" and verifies that the operator creates a load balancer and that
+// the load balancer has a private IP address.
+func TestInternalLoadBalancer(t *testing.T) {
+	cl, ns, err := getClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	infraConfig := &configv1.Infrastructure{}
+	err = cl.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig)
+	if err != nil {
+		t.Fatalf("failed to get infrastructure config: %v", err)
+	}
+
+	switch infraConfig.Status.Platform {
+	case configv1.AWSPlatformType:
+	case configv1.AzurePlatformType:
+	case configv1.BareMetalPlatformType:
+		t.Skip("test skipped on bare metal")
+	case configv1.GCPPlatformType:
+	case configv1.LibvirtPlatformType:
+		t.Skip("test skipped on libvirt")
+	case configv1.OpenStackPlatformType:
+	case configv1.NonePlatformType:
+		t.Skip("test skipped on \"none\" platform type")
+	case configv1.VSpherePlatformType:
+		t.Skip("test skipped on vSphere")
+	}
+
+	dnsConfig := &configv1.DNS{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		t.Fatalf("failed to get DNS 'cluster': %v", err)
+	}
+
+	name := types.NamespacedName{Namespace: ns, Name: ingressControllerName}
+	domain := name.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newIngressController(name.Name, name.Namespace, domain, operatorv1.LoadBalancerServiceStrategyType)
+	ic.Spec.EndpointPublishingStrategy.LoadBalancer = &operatorv1.LoadBalancerStrategy{
+		Scope: operatorv1.InternalLoadBalancer,
+	}
+	if err := cl.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller %s: %v", name, err)
+	}
+
+	// Wait for the load balancer and DNS to be ready.
+	err = waitForIngressControllerCondition(cl, 5*time.Minute, name, defaultAvailableConditions...)
+	if err != nil {
+		t.Errorf("failed to observe expected conditions: %v", err)
+	}
+
+	// Wait for the service to have at least one ingress.
+	lbService := &corev1.Service{}
+	err = wait.PollImmediate(1*time.Second, 2*time.Minute, func() (bool, error) {
+		if err := cl.Get(context.TODO(), ingresscontroller.LoadBalancerServiceName(ic), lbService); err != nil {
+			return false, nil
+		}
+		if len(lbService.Status.LoadBalancer.Ingress) == 0 {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("failed to get LoadBalancer service: %v", err)
+	} else {
+		for _, ingress := range lbService.Status.LoadBalancer.Ingress {
+			if len(ingress.IP) != 0 {
+				if !networkutils.IsPrivateAddress(ingress.IP) {
```

Comment:
Is this a safe assumption? I kind of expected you'd assert appropriate k8s metadata (i.e. annotations) since what "private" means is up to the cloud provider.

Reply:
I was hoping to test that the thing was in practice internal-only, but that may be impractical, and checking annotations would be a lot easier.  Is this check still useful as an end-to-end test, as opposed to as a unit test?

Reply:
Hmm... my thinking is that our contract with k8s in this regard is that we declare intent for internal by setting known annotations on the service. That's what the docs tell us to do in the API. It doesn't promise that the ingress IP will be internal as defined by IsPrivateAddress (or any other algorithm) or that an IP will be specified instead of a hostname. So I feel like if given our inputs we produce the output "a correctly annotated service" we can call that good enough and trust that we've satisfied our end of the contract with k8s.

If we wanted deeper assertions on the downstream effects of the annotation (e.g. an IP in a certain CIDR when using AWS), those seem platform-specific and less crucial to verifying general correctness here. What do you think?

Reply:
I think that makes sense.  And I suppose the end-to-end test is useful for ensuring the operator sets the right status in `enforceEffectiveEndpointPublishingStrategy` and that `ensureLoadBalancerService` does the right thing with that status.